storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala

```diff
@@ -444,6 +444,103 @@ final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = f
 
   override def hashCode(): Int = if (negateCharSet) charSet.hashCode() else -charSet.hashCode()
 
+  override def toString: String = {
+    def isPrintable(c: Char): Boolean =
+      ((c >= 32) && (c <= 127)) || (!Character.isISOControl(c) && {
+        val block = Character.UnicodeBlock.of(c)
+        block != null && block != Character.UnicodeBlock.SPECIALS
+      })
+
+    def singleCharToString(c: Char): String =
+      if (CharClassPattern.allSpecialChars.contains(c)) {
+        //this character needs escape
+        "\\" + c
+      } else if (CharClassPattern.revEscapes.contains(c)) {
+        "\\" + CharClassPattern.revEscapes(c)
+      } else if ((c >= 128) && (c <= 255)) {
+        "\\x%02X".format(c.toInt)
+      } else if (isPrintable(c)) {
+        c.toString
+      } else if (!c.isSurrogate && (c <= 65535)) {
+        "\\u%04X".format(c.toInt)
+      } else {
+        //we don't support surrogates anyway, but let's be able to print them
+        val surrogatePair = Character.toChars(c)
+        //one day Unicode consortium will add 3-chars surrogate, lets handle it today
+        surrogatePair.foldLeft("")((str, ch) => str + "\\u%04X".format(ch.toInt))
+      }
+
+    def unknownSetToString: String = {
+      def listToString(lst: List[Char]): String =
+        lst.size match {
+          case 0 => ""
+          case 1 => singleCharToString(lst.head)
+          case x if x > 3 => {
+            //sequence of 4 chars or more "abcd" -> "a-d"
+            val startChar = singleCharToString(lst.last)
+            val endChar = singleCharToString(lst.head)
+            s"$startChar-$endChar"
+          }
+          case _ => {
+            //sequence like "abc", no sense to convert to "a-c"
+            lst.map(c => singleCharToString(c)).mkString
+          }
+        }
+
+      val sortedSet = charSet.toList.sorted
+      val (strings, pending) = sortedSet.foldLeft((Nil: List[String], Nil: List[Char])) {
+        case ((strings, pending), nextChar) => {
+          if (pending.isEmpty) {
+            (strings, nextChar :: pending)
+          } else {
+            if (nextChar == (pending.head + 1)) {
+              //next char in a continious sequence, add to pending
+              (strings, nextChar :: pending)
+            } else {
+              (listToString(pending) :: strings, nextChar :: Nil)
+            }
+          }
+        }
+      }
+
+      val finalStrings = if (pending.isEmpty) {
+        strings
+      } else {
+        listToString(pending) :: strings
+      }
+
+      finalStrings.reverse.mkString
+    }
+
+    if (charSet.isEmpty) {
```

Comment:
One trick you can do with such if clauses is to use pattern-matching instead:
```scala
charSet.toList match {
  case Nil => "Nil"
  case h :: Nil => "Head"
  case h :: t => "Head and tail"
}
```
Maybe you'll found it more expressive.

Reply:
Thanks for the tip. Finally decided to leave as is here due to possible performance hit (Set,toList conversion). Will definitely use this feature in future.